src/appleseed/renderer/kernel/lighting/volumelightingintegrator.cpp

```diff
@@ -0,0 +1,601 @@
+
+        const ShadingContext&   m_shading_context;
+        SamplingContext&        m_sampling_context;
+    };
+}
+
+//
+// VolumeLightingIntegrator class implementation.
+//
+// Call graph:
+//
+//   compute_radiance
+//       add_single_distance_sample_contribution
+//
+//   compute_radiance_exponential_sampling_only
+//       add_single_distance_sample_contribution_exponential_only
+//
+//   add_single_distance_sample_contribution
+//       take_single_direction_sample
+//
+//   add_single_distance_sample_contribution_exponential_only
+//       take_single_direction_sample
+//
+//   take_single_direction_sample
+//       DirectLightingIntegrator::take_single_material_sample
+//       DirectLightingIntegrator::add_emitting_triangle_sample_contribution
+//       DirectLightingIntegrator::add_non_physical_light_sample_contribution
+
+VolumeLightingIntegrator::VolumeLightingIntegrator(
+    const ShadingContext&           shading_context,
+    const BackwardLightSampler&     light_sampler,
+    const Volume&                   volume,
+    const ShadingRay&               volume_ray,
+    const void*                     volume_data,
+    const ShadingPoint&             shading_point,
+    const int                       scattering_modes,
+    const size_t                    distance_sample_count,
+    const size_t                    light_sample_count,
+    const float                     low_light_threshold,
+    const bool                      indirect)
+    : m_shading_context(shading_context)
+    , m_light_sampler(light_sampler)
+    , m_volume(volume)
+    , m_volume_ray(volume_ray)
+    , m_volume_data(volume_data)
+    , m_shading_point(shading_point)
+    , m_time(volume_ray.m_time)
+    , m_scattering_modes(scattering_modes)
+    , m_light_sample_count(light_sample_count)
+    , m_distance_sample_count(distance_sample_count)
+    , m_rcp_distance_sample_count(
+        m_distance_sample_count == 0 ? 0.0f : 1.0f / m_distance_sample_count)
+    , m_low_light_threshold(low_light_threshold)
+    , m_indirect(indirect)
+{
+    precompute_mis_weights();
+}
+
+void VolumeLightingIntegrator::precompute_mis_weights()
+{
+    // Get full ray transmission.
+    Spectrum ray_transmission;
+    m_volume.evaluate_transmission(
+        m_volume_data, m_volume_ray, ray_transmission);
+
+    const Spectrum& scattering_coef = m_volume.scattering_coefficient(
+        m_volume_data, m_volume_ray);
+
+    Spectrum extinction_coef = m_volume.extinction_coefficient(
+        m_volume_data, m_volume_ray);
+
+    // Precompute MIS weights.
+    // MIS terms are:
+    //  - scattering albedo,
+    //  - one minus total ray transmittance (because we sample distance on segment),
+    //  - throughput of the entire path up to the sampled point.
+    // Reference: "Practical and Controllable Subsurface Scattering
+    // for Production Path Tracing", p. 1 [ACM 2016 Article].
+    for (size_t i = 0; i < Spectrum::size(); ++i)
```

Comment:
We've decided to switch to

    for (size_t i = 0, e = Spectrum::size(); i < e; ++i)

for such loops.

Change in other places.